### PR TITLE
Delete pending registrations automatically #4225

### PR DESCRIPTION
**Pull request by MasterHans**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

**Other information:**

1) Added **DeleteAll** button which has two actions depending on does any selected rows in the grid exists. First simply deletes all records in the **Invite** model. And second, delete selected rows in the grid. If the user selects at least one row in the grid the button text description changes from **"Delete All"** to **"Delete selected invitations"**.
2) Added **checkboxColumn** in grid
3) Created **humhub.admin.pending.registration.js** module. Put there all handling of deleting selected rows.

### CODE REVIEW

File: protected/humhub/modules/admin/resources/js/humhub.admin.pending.registrations.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                result = true;
+                return false;
+            }
+        });
+        return result;
+    }
+
+    var init = function (isPjax) {
+
+        $("input").change(function () {
```
Comment (buddh4) [inline protected/humhub/modules/admin/resources/js/humhub.admin.pending.registrations.js:17]: Note, once loaded a module will be initialized on every following pjax load. In the init function you should make sure you are residing in the right view, either by checking the existence of certain dom elements (selectors) or checking the [view state](https://github.com/humhub/humhub/blob/master/static/js/humhub/humhub.ui.view.js#L167). In this case you could also consider implementing a [JsWidget](https://docs.humhub.org/docs/develop/javascript-components#widgets)

Also its never a good idea to add such handlers to all `input` elements on the page, you should be more specific about which elements you want to listen to (e.g. only within a specific root element).

For example when implementing a JsWidget on the root element of the form, you could just call `this.$.find('input')`.

File: protected/humhub/modules/admin/views/pending-registrations/index.php [hunk 2]
```diff
@@ -20,6 +24,9 @@
     </div>
 
     <div class="pull-right">
+        <?php if ($dataProvider->totalCount > 0): ?>
+            <?= Html::a(Yii::t('AdminModule.user', 'Delete All'), Url::toRoute(['/admin/pending-registrations/delete-all']), ['class' => 'btn btn-danger btn-delete']); ?>
```
Comment (buddh4) [inline protected/humhub/modules/admin/views/pending-registrations/index.php:28]: Rather use a js action with confirm e.g. `Button::danger('Yii::t('AdminModule.user', 'Delete All')')->action('admin.pending.registrations.deleteAll', Url::toRoute(['/admin/pending-registrations/delete-all'])->confirm()`

File: protected/humhub/modules/admin/resources/js/humhub.admin.pending.registrations.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            }
+        });
+
+        $(".btn-delete").on("click", function (e) {
+            var keys = $("#grid").yiiGridView("getSelectedRows");
+            if (keys.length > 0) {
+                e.preventDefault();
+                if(confirm("Delete selected invitations?")){
+                    $.ajax({
+                        url: "/index.php?r=admin%2Fpending-registrations%2Fdelete-all",
```
Comment (buddh4) [inline protected/humhub/modules/admin/resources/js/humhub.admin.pending.registrations.js:33]: Do not use static urls in javascript, this will fail for example when using url rewriting rules. Check https://docs.humhub.org/docs/develop/javascript-index#module-configuration for how to transfer such data to js as module configuration or when using a JsWidget, you can use the widget options which are provided as `data-*` attributes on the widget root.

File: protected/humhub/modules/admin/views/pending-registrations/index.php [hunk 4]
```diff
@@ -6,6 +6,10 @@
 use yii\helpers\Url;
 
 /** @var $searchModel \humhub\modules\admin\models\PendingRegistrationSearch */
+/** @var $dataProvider yii\data\ActiveDataProvider */
+/** @var $types array */
+
+\humhub\modules\admin\assets\AdminPendingRegistrationsAsset::register($this);
```
Comment (buddh4) [inline protected/humhub/modules/admin/views/pending-registrations/index.php:12]: Prefer importing the namepsace.

**Review by buddh4 — CHANGES_REQUESTED** [review]

File: protected/humhub/modules/admin/assets/AdminPendingRegistrationsAsset.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    ];
+    public $sourcePath = '@admin/resources';
+
+    /**
+     * @inheritdoc
+     */
+    public $js = [
+        'js/humhub.admin.pending.registrations.js'
+    ];
+
```
Comment (buddh4) [inline protected/humhub/modules/admin/assets/AdminPendingRegistrationsAsset.php:30]: You could register the url here e.g:

```
public static function register($view)
    {
        $view->registerJsConfig('admin.pending.registrations', [
            'deleteAllUrl' => Url::to('...')
        ]);

        return parent::register($view); // TODO: Change the autogenerated stub
    }
```

In you js module call:

```
var url = module.config.deleteAllUrl;
```

File: protected/humhub/modules/admin/resources/js/humhub.admin.pending.registrations.js (hunk 1 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/resources/js/humhub.admin.pending.registrations.js:17]: Done!

File: protected/humhub/modules/admin/views/pending-registrations/index.php (hunk 2 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/views/pending-registrations/index.php:28]: Done!

File: protected/humhub/modules/admin/views/pending-registrations/index.php (hunk 4 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/views/pending-registrations/index.php:12]: Done!

File: protected/humhub/modules/admin/assets/AdminPendingRegistrationsAsset.php (hunk 5 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/assets/AdminPendingRegistrationsAsset.php:30]: Done with JsWidget

File: protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    PendingRegistrations.prototype.init = function() {
+        urlDeleteSelected = this.options.urlDeleteSelected;
+        this.$.find("input").change(function () {
+
+            var $checkBoxes = $('.regular-checkbox');
+
+            if (hasChecked($checkBoxes)) {
+                $('.delete-all').html('Delete selected rows')
+                $('.delete-all').attr('data-action-click','admin.PendingRegistrations.deleteAllSelected');
+                $('.delete-all').attr('data-action-click-url','/index.php?r=admin%2Fpending-registrations%2Fdelete-all-selected');
```
Comment (luke-) [inline protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js:54]: You always need to use `Url::to()` otherwise URL Rewriting is not supported.

File: protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    PendingRegistrations.prototype.init = function() {
+        urlDeleteSelected = this.options.urlDeleteSelected;
+        this.$.find("input").change(function () {
+
+            var $checkBoxes = $('.regular-checkbox');
+
+            if (hasChecked($checkBoxes)) {
+                $('.delete-all').html('Delete selected rows')
```
Comment (luke-) [inline protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js:52]: Missing `Yii::t`

File: protected/humhub/modules/admin/controllers/PendingRegistrationsController.php [hunk 8]
```diff
@@ -125,6 +125,51 @@ public function actionDelete($id)
...
+        if (Yii::$app->request->isPost) {
+               Invite::deleteAll();
+            $this->view->success(Yii::t(
+                'AdminModule.user',
+                'Deleted all invitations'
+            ));
+        }
+        return $this->redirect(['index']);
+    }
+    /**
```
Comment (luke-) [inline protected/humhub/modules/admin/controllers/PendingRegistrationsController.php:147]: \n

File: protected/humhub/modules/admin/assets/AdminPendingRegistrationsAsset.php [hunk 9]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+/**
+ * @link https://www.humhub.org/
+ * @copyright Copyright (c) 2017 HumHub GmbH & Co. KG
```
Comment (luke-) [inline protected/humhub/modules/admin/assets/AdminPendingRegistrationsAsset.php:5]: 2020

**Review by luke- — CHANGES_REQUESTED** [review]
I've also added some findings.

If I change the browser scaling slightly, e.g. to 110%, it looks quite strange:
image

**Comment by luke-** [comment]
We should also automatically delete invites that have not yet been registered after 90 days.
E.g. similar as here: https://github.com/humhub/humhub/blob/master/protected/humhub/modules/admin/Events.php#L53

File: protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js (hunk 7 above)
Comment (buddh4) [reply protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js:52]: See https://docs.humhub.org/docs/develop/javascript-index#module-texts

File: protected/humhub/modules/admin/assets/AdminPendingRegistrationsAsset.php [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+ * @license https://www.humhub.com/licences
+ */
+
+namespace humhub\modules\admin\assets;
+
+use yii\web\AssetBundle;
```
Comment (buddh4) [inline protected/humhub/modules/admin/assets/AdminPendingRegistrationsAsset.php:11]: Use the `humhub\components\assets\AssetBundle` instead

File: protected/humhub/modules/admin/controllers/PendingRegistrationsController.php [hunk 11]
```diff
@@ -125,6 +125,51 @@ public function actionDelete($id)
...
+        if (Yii::$app->request->isPost) {
+            $ids = Yii::$app->request->post('id');
+            if (!empty($ids)) {
+                foreach ($ids as $id) {
+                    $invitation = Invite::findOne(['id' => $id]);
+                    $invitation->delete();
+                }
+                $this->view->success(Yii::t(
+                    'AdminModule.user',
+                    'Deleted all selected invitations'
```
Comment (buddh4) [inline protected/humhub/modules/admin/controllers/PendingRegistrationsController.php:166]: Selected invitations have been deleted

File: protected/humhub/modules/admin/controllers/PendingRegistrationsController.php [hunk 12]
```diff
@@ -125,6 +125,51 @@ public function actionDelete($id)
...
+     * @throws HttpException
+     * @throws \Throwable
+     */
+    public function actionDeleteAll()
+    {
+        if (Yii::$app->request->isPost) {
+               Invite::deleteAll();
+            $this->view->success(Yii::t(
+                'AdminModule.user',
+                'Deleted all invitations'
```
Comment (buddh4) [inline protected/humhub/modules/admin/controllers/PendingRegistrationsController.php:142]: All invitations have been deleted

File: protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js [hunk 13]
```diff
@@ -0,0 +1,65 @@
+humhub.module('admin.PendingRegistrations', function(module, require, $) {
+    var object = require('util').object;
+    var Widget = require('ui.widget').Widget;
+    var client = require('client');
+    var urlDeleteSelected;
+
+    var PendingRegistrations = function(node, options) {
```
Comment (buddh4) [inline protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js:7]: Just use:

```
var PendingRegistrations = Widget.extend();
```

and remove the following line:

```
object.inherits(PendingRegistrations, Widget);
```

File: protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+                url: urlDeleteSelected,
+                type: "POST",
+                data: {id: keys},
+            })
+        }).catch(function(e) {
+            module.log.error(e, true);
+        })
+    };
+
+    PendingRegistrations.deleteAll = function(evt) {
```
Comment (buddh4) [inline protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js:26]: Change functions to:

`PendingRegistrations.prototype.deleteAll` and `PendingRegistrations.prototype.deleteAllSelected` otherwise those function won't be bound to your widget instance and are only available as static functions, also you won't be able to access `this`.

Then change the action buttons to:

`->action('deleteAll', Url::toRoute(['/admin/pending-registrations/delete-all']))`

and

  $('.delete-all').attr('data-action-click','deleteAllSelected');`

As long as those buttons reside in your widget, the widget function will be called.

File: protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+        client.post(evt).then(function() {
+            that.reload();
```
Comment (buddh4) [inline protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js:28]: That is currently not initialized. Add `var that = this` prior to the `client.post()` call.

File: protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+            }
+        });
+        return result;
+    }
+
+    PendingRegistrations.prototype.init = function() {
+        urlDeleteSelected = this.options.urlDeleteSelected;
+        this.$.find("input").change(function () {
+
+            var $checkBoxes = $('.regular-checkbox');
```
Comment (buddh4) [inline protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js:49]: I think `var $selection = that.$.find(':checked')` should to the trick and we can remove the `hasChecked()` function then just:

```
if($selection.length) {
 // Do stuff
}
```

(of course you'll have to add the `var that = this` prior to ` this.$.find("input").change(function () {`)

File: protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+            if ($(this).prop("checked")) {
+                result = true;
+                return false;
+            }
+        });
+        return result;
+    }
+
+    PendingRegistrations.prototype.init = function() {
+        urlDeleteSelected = this.options.urlDeleteSelected;
```
Comment (buddh4) [inline protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js:46]: I'd rather directly check `this.options.urlDeleteSelected` in the `deleteAllSelected ` instead of adding this static variable.

File: protected/humhub/modules/admin/widgets/views/PendingRegistrationsWidgetView.php [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+                return isset($types[$data->source]) ?: Html::encode($data->source);
+            },
+        ],
+        [
+            'header' => Yii::t('AdminModule.user', 'Actions'),
+            'class' => 'yii\grid\ActionColumn',
+            'template' => '{resend} {delete}',
+            'buttons' => [
+                'resend' => function ($url, $model, $key) {
+                    return Html::a(
```
Comment (buddh4) [inline protected/humhub/modules/admin/widgets/views/PendingRegistrationsWidgetView.php:70]: `Button::primary()->link(Url::to(['resend', 'id' => $model->id]))->icon('envelope')->xs()` we don't need the `tt` class since we don't have a tooltip.

File: protected/humhub/modules/admin/widgets/views/PendingRegistrationsWidgetView.php [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+                        '<i class="fa fa-envelope"></i>',
+                        Url::to(['resend', 'id' => $model->id]),
+                        ['class' => 'btn btn-primary btn-xs tt']
+                    );
+                },
+                'delete' => function ($url, $model, $key) {
+                    return Html::a(
```
Comment (buddh4) [inline protected/humhub/modules/admin/widgets/views/PendingRegistrationsWidgetView.php:77]: `Button::primary()->link(Url::to(['delete', 'id' => $model->id]))->icon('delete')->xs()` we don't need the `tt` class since we don't have a tooltip.

**Review by buddh4 — CHANGES_REQUESTED** [review]

File: protected/humhub/modules/admin/widgets/views/PendingRegistrationsWidgetView.php (hunk 19 above)
Comment (buddh4) [reply protected/humhub/modules/admin/widgets/views/PendingRegistrationsWidgetView.php:77]: I think those two functions should be `post` links instead.

File: protected/humhub/modules/admin/resources/js/humhub.admin.pending.registrations.js (hunk 3 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/resources/js/humhub.admin.pending.registrations.js]: Done with Widget!

File: protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js (hunk 7 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js:52]: Done!

File: protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js (hunk 6 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js:54]: Done!

Comment (MasterHans) [reply protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js]: > I've also added some findings.
> 
> If I change the browser scaling slightly, e.g. to 110%, it looks quite strange:
> image

Done!

File: protected/humhub/modules/admin/controllers/PendingRegistrationsController.php (hunk 8 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/controllers/PendingRegistrationsController.php:147]: @luke- I am sorry did not catch what does it mean...

File: protected/humhub/modules/admin/assets/AdminPendingRegistrationsAsset.php (hunk 9 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/assets/AdminPendingRegistrationsAsset.php:5]: done!

File: protected/humhub/modules/admin/assets/AdminPendingRegistrationsAsset.php (hunk 10 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/assets/AdminPendingRegistrationsAsset.php:11]: Done!

File: protected/humhub/modules/admin/controllers/PendingRegistrationsController.php (hunk 11 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/controllers/PendingRegistrationsController.php:166]: Done!

File: protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js (hunk 13 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js]: done!

File: protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js (hunk 14 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js]: Done! Thank you!

File: protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js (hunk 15 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js:32]: Done with client.reload()
Redirect in controller also works, but I did not want to leave empty  
client.post(evt).then(function()  like this:

```
        client.post(evt).then(function () {

        }).catch(function (e) {
            module.log.error(e, true);
        })
```


)

File: protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js (hunk 16 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js:57]: Done!

File: protected/humhub/modules/admin/widgets/views/PendingRegistrationsWidgetView.php (hunk 18 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/widgets/views/PendingRegistrationsWidgetView.php:80]: Done!

File: protected/humhub/modules/admin/widgets/views/PendingRegistrationsWidgetView.php (hunk 19 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/widgets/views/PendingRegistrationsWidgetView.php:87]: Done!

File: protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js (hunk 15 above)
Comment (buddh4) [reply protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js]: This would have been an alternative:

```
client.post(evt)..catch(function (e) {
    module.log.error(e, true);
});
```

In this case you could just omit the `then` handler.

File: protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js (hunk 17 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js]: Done!

File: protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js (hunk 15 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/resources/js/humhub.admin.PendingRegistrations.js]: Done!

File: protected/humhub/modules/admin/controllers/PendingRegistrationsController.php (hunk 8 above)
Comment (luke-) [reply protected/humhub/modules/admin/controllers/PendingRegistrationsController.php:149]: @MasterHans There is a missing newline between the end of the method and the next one.

**Comment by luke-** [comment]
Looks quite good for me! Can you also implement an automatic cleanup after 90 days, similar to the logfile cleanup.

Would be great if the auto delete time could be configured as module variable, e.g.:
https://github.com/humhub/humhub/blob/master/protected/humhub/modules/admin/Module.php#L56

**Comment by MasterHans** [comment]
> We should also automatically delete invites that have not yet been registered after 90 days.
> E.g. similar as here: https://github.com/humhub/humhub/blob/master/protected/humhub/modules/admin/Events.php#L53

Done!

File: protected/humhub/modules/admin/jobs/CleanupPendingRegistrations.php [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * @since 1.8
+ */
+
+class CleanupPendingRegistrations extends ActiveJob
+{
+    /**
+     * @var int seconds before delete old pending registrations messages
+     */
+    public $cleanupInterval = 60 * 60 * 24 * 90;
```
Comment (ArchBlood) [inline protected/humhub/modules/admin/jobs/CleanupPendingRegistrations.php:26]: @luke- question about the interval, is this related to Cron? Then wouldn't `60 * 60 * 24 * 7` be for the best? 
### See;
https://github.com/humhub/humhub/blob/a5da4c1b0b9eae3c38345f46abfa93ab8d030d42/protected/humhub/modules/admin/jobs/CleanupLog.php#L23-L26

Although I'm thinking something like `59 59 23 * * 7 *` would be better? As some Cron tasks don't calculate the same way :thinking:

File: protected/humhub/modules/admin/controllers/PendingRegistrationsController.php (hunk 8 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/controllers/PendingRegistrationsController.php:150]: Done!

**Comment by MasterHans** [comment]
> We should also automatically delete invites that have not yet been registered after 90 days.
> E.g. similar as here: https://github.com/humhub/humhub/blob/master/protected/humhub/modules/admin/Events.php#L53

Done!

**Comment by MasterHans** [comment]
> We should also automatically delete invites that have not yet been registered after 90 days.
> E.g. similar as here: https://github.com/humhub/humhub/blob/master/protected/humhub/modules/admin/Events.php#L53

I got it about variable, but
I am sorry I did not catch  about  automatically delete invites - I have already implemented your previous task :
 
> > We should also automatically delete invites that have not yet been registered after 90 days.
> > E.g. similar as here: https://github.com/humhub/humhub/blob/master/protected/humhub/modules/admin/Events.php#L53
>

**Comment by MasterHans** [comment]
> Would be great if the auto delete time could be configured as module variable, e.g.:
> https://github.com/humhub/humhub/blob/master/protected/humhub/modules/admin/Module.php#L56

Done!

**Comment by luke-** [comment]
There is should be no difference. You had only created an additional feature to delete invites with checkboxes.

**Comment by MasterHans** [comment]
> There is should be no difference. You had only created an additional feature to delete invites with checkboxes.
I am sorry but I had created this:
Did you saw this commit
image
Please check out this commit

**Review by luke- — APPROVED** [review]

File: protected/humhub/modules/admin/jobs/CleanupPendingRegistrations.php (hunk 20 above)
Comment (ArchBlood) [reply protected/humhub/modules/admin/jobs/CleanupPendingRegistrations.php]: When commenting on this, I was wondering if this would work for all systems if it were related to cron jobs :thinking:
